### ISSUE TITLE
bionic is out! Time to rebuild everything!

### ISSUE THREAD

**Issue by asottile** [deadsnakes/issues#62]

**Comment by asottile** [comment]
Interesting, bionic also ships with a `python3.7` package -- I guess I'll leave that one out for now.

**Comment by asottile** [comment]
hmmm, rebuilding the early pythons is proving difficult.  Seems the line endings got messed up when moving to git? CC @fkrull -- I'm going to try and manually restore them

**Comment by asottile** [comment]
a little `xargs unix2dos` and they're building fine now -- I separated the line ending changes into a different commit

though I imagine I won't hit this going forward

**Comment by asottile** [comment]
[ ] python2.3
[ ] python2.4
[ ] python2.5
[ ] python2.6
[ ] python3.1
[ ] python3.2
[ ] python3.3
[ ] python3.4
[ ] python3.5

**Comment by asottile** [comment]
2.3, 2.4, and 2.5 each have this same error -- will have to investigate further:

```
make[1]: Entering directory '/tmp/code/Doc'
TEXINPUTS=/tmp/code/Doc/commontex: /tmp/code/build-static/python /tmp/code/Doc/tools/mkhowto --html --about html/stdabout.dat --iconserver ../icons --favicon ../icons/pyfav.png --address "See <i><a href=\"about.html\">About this document...</a></i> for information on suggesting changes." --up-link ../index.html --up-title "Python Documentation Index" --global-module-index "../modindex.html" --dvips-safe --dir html/api api/api.tex
*** Session transcript and error messages are in /tmp/code/Doc/html/api/api.how.
*** Exited with status 2.
The relevant lines from the transcript are:
------------------------------------------------------------------------
+++ perl /tmp/code/Doc/tools/node2label.pl *.html
Can't locate labels.pl in @INC (@INC contains: /tmp/code/Doc/perl /etc/perl /usr/local/lib/x86_64-linux-gnu/perl/5.26.1 /usr/local/share/perl/5.26.1 /usr/lib/x86_64-linux-gnu/perl5/5.26 /usr/share/perl5 /usr/lib/x86_64-linux-gnu/perl/5.26 /usr/share/perl/5.26 /usr/local/lib/site_perl /usr/lib/x86_64-linux-gnu/perl-base) at /tmp/code/Doc/tools/node2label.pl line 12.
*** Session transcript and error messages are in /tmp/code/Doc/html/api/api.how.
*** Exited with status 2.
+++ TEXINPUTS=/tmp/code/Doc/api:/tmp/code/Doc/commontex:/tmp/code/Doc/paper-letter:/tmp/code/Doc/texinputs:
+++ latex api
+++ latex2html -init_file api.l2h -dir /tmp/code/Doc/html/api /tmp/code/Doc/api/api.tex
+++ perl /tmp/code/Doc/tools/node2label.pl *.html
Makefile:378: recipe for target 'html/api/api.html' failed
```

**Comment by asottile** [comment]
hmm, the `labels.pl` appears to be a make race (?), it's a file provided by python itself

**Comment by asottile** [comment]
hmm nope, still failing -- I'll poke at this tomorrow /me shakes fists at perl

**Comment by asottile** [comment]
This appears to be the difference for `labels.pl`:

```
(xenial) $ perl -e "print \"@INC\"" | xargs -n1
/etc/perl
/usr/local/lib/x86_64-linux-gnu/perl/5.22.1
/usr/local/share/perl/5.22.1
/usr/lib/x86_64-linux-gnu/perl5/5.22
/usr/share/perl5
/usr/lib/x86_64-linux-gnu/perl/5.22
/usr/share/perl/5.22
/usr/local/lib/site_perl
/usr/lib/x86_64-linux-gnu/perl-base
.
```

```
(bionic) # perl -e "print \"@INC\"" | xargs -n1        
/etc/perl
/usr/local/lib/x86_64-linux-gnu/perl/5.26.1
/usr/local/share/perl/5.26.1
/usr/lib/x86_64-linux-gnu/perl5/5.26
/usr/share/perl5
/usr/lib/x86_64-linux-gnu/perl/5.26
/usr/share/perl/5.26
/usr/local/lib/site_perl
/usr/lib/x86_64-linux-gnu/perl-base
```

(the hard-to-spot difference being `.`)

**Comment by asottile** [comment]
Here's the changelog from perl's side: https://perldoc.perl.org/perl5260delta.html#Removal-of-the-current-directory-(%22.%22)-from-%40INC

**Comment by asottile** [comment]
3.1, 3.2, and 3.3 appear to be failing in this way: 

```
	`cd debian/tmp; for i in _bisect _collections _elementtree _functools _hashlib _heapq _locale _pickle _random _socket _ssl _stringio _struct array atexit binascii cmath datetime fcntl grp itertools math operator pyexpat select spwd syslog time unicodedata zipimport zlib; do \
		test -e usr/lib/python3.1/lib-dynload/$i*.so \
		  && echo usr/lib/python3.1/lib-dynload/$i*.so; \
	  done`
ls -l  debian/python3.1-minimal/usr/lib/python3.1/lib-dynload/*.so
ls: cannot access 'debian/python3.1-minimal/usr/lib/python3.1/lib-dynload/*.so': No such file or directory
```

**Comment by fkrull** [comment]
You people and your time zones. ;) I'll take a look later, when I'm not at work.

**Comment by asottile** [comment]
Makes a lot of sense, on a related topic -- what would you suggest for #63 (upstream dropped trusty's openssl 1.0.1)

**Comment by asottile** [comment]
Looks like python3.5 has the patch, but I'll also need to fix the ssl bits for python2.6 and python3.4

**Comment by asottile** [comment]
`socket.ssl` works in xenial, so I'll see what I can do to get those working too

**Comment by fkrull** [comment]
There's a patch for 3.4 [in this issue](https://bugs.python.org/issue30008). It applies cleanly, but I'll have to see if it builds and works correctly. This seems like a smaller patch than the other one so it might be easier to backport.

That issue also has a patch for 2.7 which might be useful for 2.x ("might" because I'm pretty sure the ssl module was heavily rebuilt during 2.7's lifetime).

**Comment by fkrull** [comment]
...alternatively, I could stop and search for a second and learn that bionic has an [openssl1.0 source package](https://packages.ubuntu.com/source/bionic/openssl1.0). I don't know how long-term of a solution that is, but -- assuming it works -- it's certainly the easiest.

**Comment by asottile** [comment]
oh hey yeah that is probably the easiest -- good find!

**Comment by fkrull** [comment]
Ok! I changed all the dependencies from `libssl-dev` to `libssl1.0-dev`. Everything worked from there, so the other problems were caused by the missing _ssl module as well. They all import fine now.

I'm not sure how long-term of a solution this is going to be. If the libssl1.0.0 package goes away, the issue'll come back. But that's not going to be relevant for another two years, so this is for Future Us to worry about.
